src/lib/reasoners/satml.ml

```diff
@@ -614,14 +594,22 @@ module Make (Th : Theory.S) : SAT_ML with type th = Th.t = struct
       Printer.print_dbg
         "[satml] cancel until %d (current is %d)@." lvl (decision_level env);
     cancel_ff_lvls_until env lvl;
-    let repush = ref [] in
     if decision_level env > lvl then begin
-      env.qhead <- Vec.get env.trail_lim lvl;
-      for c = Vec.size env.trail - 1 downto env.qhead do
+      let head = ref (Vec.get env.trail_lim lvl) in
+      env.qhead <- !head;
+      for c = !head to Vec.size env.trail - 1 do
         let a = Vec.get env.trail c in
         if Options.get_minimal_bj () && a.var.level <= lvl then begin
+          assert (a.var.level >= 0);
           assert (a.var.level = 0 || a.var.reason != None);
-          repush := a :: !repush
+          (* bclement: Ported over from the old [enqueue_assigned] -- not sure
+             what this actually does *)
```

Comment:
Actually I understood this and forgot: the `timp` field is set to `1` when the atom is learned using theory propagation (through `th_entailed`), and we then set it to `-1` when we backtrack at a lower level so that we can check whether it is still implied at an earlier level or we need to propagate it back to the theory (see `theory_propagate`). I will update the comment.

We should probably replace this with a tri-state field :)